### PR TITLE
add async + distopt to sft

### PR DESCRIPTION
**Pull request by MaximumEntropy**
# What does this PR do ?

Add a one line overview of what this PR aims to accomplish.

**Collection**: NLP

# Changelog 
- Add specific line by line info of high level changes in this PR.

# Usage
* You can potentially add a usage example below

```python
# Add a code snippet demonstrating how to use this 
```

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [ ] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

**Review by aklife97 — APPROVED** [review]
LGTM, thank you!

File: nemo/collections/nlp/models/language_modeling/megatron_gpt_sft_model.py [hunk 1]
```diff
@@ -296,6 +296,15 @@ def fwd_bwd_step(self, dataloader_iter, batch_idx, forward_only):
         tensor_shape = [seq_length, get_micro_batch_size(), self.cfg.hidden_size]
         data_iter = get_iterator_k_split(batch, get_num_microbatches())
 
+        # handle asynchronous grad reduction
+        no_sync_func = None
+        grad_sync_func = None
+        param_sync_func = None
+        if not forward_only and self.with_distributed_adam:
+            no_sync_func = partial(self._optimizer.no_sync, greedy_grad_copy=self.megatron_amp_o2,)
```
Comment (hsiehjackson) [inline nemo/collections/nlp/models/language_modeling/megatron_gpt_sft_model.py:304]: @MaximumEntropy Looks like we forgot `from functools import partial`

Comment (MaximumEntropy) [reply nemo/collections/nlp/models/language_modeling/megatron_gpt_sft_model.py:304]: Ooops can you approve this? https://github.com/NVIDIA/NeMo/pull/7026

Comment (aklife97) [reply nemo/collections/nlp/models/language_modeling/megatron_gpt_sft_model.py:304]: nice catch! approved the import PR

**Pull request merged**